perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -5021,16 +5028,20 @@ public void recalculateMemberGroupStatusRecursively(PerunSession sess, Member me
 		if (saveStatuses) {
 			groupsManagerImpl.setIndirectGroupStatus(sess, member, group, newStatus);
 
-			if (newStatus.equals(MemberGroupStatus.EXPIRED)) {
-				getPerunBl().getAuditer().log(sess, new MemberExpiredInGroup(member, group));
-			} else if (newStatus.equals(MemberGroupStatus.VALID)) {
-				getPerunBl().getAuditer().log(sess, new MemberValidatedInGroup(member, group));
+			if (!newStatus.equals(previousStatus)) {
+				if (newStatus.equals(MemberGroupStatus.EXPIRED)) {
+					getPerunBl().getAuditer().log(sess, new MemberExpiredInGroup(member, group));
+				} else if (newStatus.equals(MemberGroupStatus.VALID)) {
+					getPerunBl().getAuditer().log(sess, new MemberValidatedInGroup(member, group));
+				}
 			}
 		}
 
 		// check recursively all parent groups
 		for (Group affectedGroup : affectedGroups) {
-			recalculateMemberGroupStatusRecursively(sess, member, affectedGroup);
+			MemberGroupStatus previousStatusInAffectedGroup = getDirectMemberGroupStatus(sess, member, affectedGroup);
```

Comment:
I am not sure, but I think that `previousStatusInAffectedGroup` should be calculated as `getTotalMemberGroupStatus` before `setIndirectGroupStatus` at line 5029, because we care about member's overall status in the group, not just his direct status. What do you think, @metodej @Johaney-s?

Reply:
Yes, you are correct, it should, thanks! Fixed.

Reply:
I think the previous state should be calculated before `setIndirectGroupStatus` because that's where the status might change. Also, it is worth thinking about calculating the previous state similar way in methods `moveGroup`, `createGroupUnion` and `removeGroupUnion`.

Reply:
Good point, It's fixed. Thanks.